### PR TITLE
Add capability for the v1/connect/ca/roots endpoint to return a PEM encoded certificate chain

### PR DESCRIPTION
**Pull request by mkeeler**
The `v1/connect/ca/roots` endpoint has picked up a `pem=true` argument. When set this causes the endpoint to output data with a `Content-Type` set to `application/pem-certificate-chain` and the response body to be a PEM encoded certificate chain of all root and intermediate certs that would normally be contained in the response.

This output format can be useful in situations where you need a certificate bundle for passing off to another application such as the Consul CLI when they will interact with an API presenting behind a Connect CA signed TLS cert (such as when `auto_encrypt` or `auto_config` are used for generating a Consul clients HTTPs certificate).

### CODE REVIEW

File: agent/connect_ca_endpoint.go [hunk 1]
```diff
@@ -15,13 +16,35 @@ func (s *HTTPHandlers) ConnectCARoots(resp http.ResponseWriter, req *http.Reques
...
 		return nil, err
 	}
 
-	return reply, nil
+	if !pemResponse {
+		return reply, nil
+	}
+
+	// defined in RFC 8555 and registered with the IANA
+	resp.Header().Set("Content-Type", "application/pem-certificate-chain")
```
Comment (i0rek) [inline agent/connect_ca_endpoint.go:39]: TIL 🤩

**Review by i0rek — APPROVED** [review]
LGTM

File: agent/connect_ca_endpoint.go (hunk 1 above)
Comment (mkeeler) [reply agent/connect_ca_endpoint.go:39]: This is the authoritative list of standard mime types

https://www.iana.org/assignments/media-types/media-types.xhtml

File: website/pages/api-docs/connect/ca.mdx [hunk 2]
```diff
@@ -31,6 +31,13 @@ The table below shows this endpoint's support for
 | ---------------- | ----------------- | ------------- | ------------ |
 | `YES`            | `all`             | `none`        | `none`       |
 
+### Paramaters
```
Comment (rboyer) [inline website/pages/api-docs/connect/ca.mdx:34]:
```suggestion
### Parameters
```

File: agent/connect_ca_endpoint.go [hunk 3]
```diff
@@ -15,13 +16,35 @@ func (s *HTTPHandlers) ConnectCARoots(resp http.ResponseWriter, req *http.Reques
...
+	for _, root := range reply.Roots {
+		resp.Write([]byte(root.RootCert))
```
Comment (rboyer) [inline agent/connect_ca_endpoint.go:41]: Should you do something with the errors returned by `Write` ?

**Review by rboyer — CHANGES_REQUESTED** [review]
Found some minor things.

File: website/pages/api-docs/connect/ca.mdx [hunk 4]
```diff
@@ -31,6 +31,13 @@ The table below shows this endpoint's support for
...
+
+- `pem` `(boolean: false)` - Specifies that the return body should be a PEM encoded
+  certificate chain suitable for use by applications needing to trust Connect CA
+  signed certificates. The Content-Type will be set to `application/pem-certificate-chain`
+  to indcate the format of the response.
```
Comment (blake) [inline website/pages/api-docs/connect/ca.mdx:39]:
```suggestion
  to indicate the format of the response.
```

File: agent/connect_ca_endpoint.go (hunk 3 above)
Comment (mkeeler) [reply agent/connect_ca_endpoint.go]: Why not. I changed it to just return the error from the function. There isn't anything we can do to recover really nor does it seem important to try.

File: agent/connect_ca_endpoint.go [hunk 5]
```diff
@@ -15,13 +16,39 @@ func (s *HTTPHandlers) ConnectCARoots(resp http.ResponseWriter, req *http.Reques
 		return nil, nil
 	}
 
+	pemResponse := false
+	if pemParam := req.URL.Query().Get("pem"); pemParam != "" {
+		val, err := strconv.ParseBool(pemParam)
+		if err != nil {
+			return nil, BadRequestError{Reason: "The 'pem' query paramter must be a boolean value"}
```
Comment (rboyer) [inline agent/connect_ca_endpoint.go:23]:
```suggestion
			return nil, BadRequestError{Reason: "The 'pem' query parameter must be a boolean value"}
```

**Review by rboyer — APPROVED** [review]
LGTM, just one more typo.

File: agent/connect_ca_endpoint.go (hunk 5 above)
Comment (mkeeler) [reply agent/connect_ca_endpoint.go:23]: My spelling was apparently even more terrible this week than most.

**Review by rboyer — APPROVED** [review]
:shipit:

**Pull request merged**